---
title: "Proteomic Ovarian Cancer Network Analysis"
author: "Christopher Conley, Pei Wang, Umut Ozbek, Jie Peng"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
---

```{r, echo = FALSE}
library(knitr)
opts_chunk$set(message=F, warning=F)
```

## Input

Load the gene coordinate annotations for protein expressions and genomic CNA.

```{r}
library(Biobase)
yinfo <- pData(featureData(readRDS(file ="~/scratch-data/neta-poc/nohc-mad-filtered/poc-prot-eset-dropout-std.rds")))
yinfo$chr <- paste0("chr", yinfo$chr)
xinfo <- pData(featureData(readRDS(file = "~/scratch-data/neta-poc/nohc-mad-filtered/poc-cna-eset-dropout-std.rds")))
```

Load the *Boot.Vote* CNA-protein network. 

```{r}
library(Matrix)
bv <- readRDS(file = "~/scratch-data/neta-poc/nohc-mad-boot-vote/03/poc-boot-vote-p100-B1000.rds")
net <- bv$bv
rownames(net$xy) <- xinfo$id; colnames(net$xy) <- yinfo$id;
rownames(net$yy) <- yinfo$id; colnames(net$yy) <- yinfo$id;
```

Load the degree distributions for the ensemble of bootstrapped networks. 

```{r}
bdeg <- bv$bdeg
colnames(bdeg$xy) <- xinfo$id; colnames(bdeg$yy) <- yinfo$id;
```

Load the Gene Ontology mappings from Org.Hs.eg.db for enrichment analysis. 

```{r}
go2eg <- readRDS("~/scratch-data/neta-poc/nohc-mad-filtered/prot-go-bp-to-entrez-gene-list.rds")
library(AnnotationDbi)
#human readable 
process_alias <- Term(names(go2eg))
```

## MSigDB Collections 

Import gene set collections from MSigDB. Details on the collections are available at the [MSigDB website](software.broadinstitute.org/gsea/msigdb/collection_details.jsp). 

```{r}
msig_import <- function(file, GO = F) { 
  p2r77map2 <- readRDS(file = "~/scratch-data/neta-poc/nohc-mad-filtered/prot-refseq-to-entrezgene-map.rds")
  c1 <- qusage::read.gmt(file)
  #switch to refseq id because of graph id and to include all isoforms
  c1 <- lapply(c1, function(x) as.character(p2r77map2$refseq[ p2r77map2$entrezgene %in% x ]))
  #remove duplicate entrez genes
  c1 <- lapply(c1, function(x) x[!duplicated(x)])
  c1_size <- sapply(c1, length)
  if (GO) { 
    c1 <- c1[ c1_size >= 15 & c1_size <= 300]
  }
  c1 <- c1[ c1_size >= 15]
  #assure each go term has a list of non-duplicate ids
  stopifnot(all(sapply(c1, anyDuplicated) == 0))
  c1
}
```


### C5 GO Biological Process


```{r}
fc5 <-  "~/scratch-data/neta-poc/MSigDB/c5.bp.v6.0.entrez.gmt"
c5 <- msig_import(fc5, GO = T)
#summary(c1)
```

### C1 Positional gene sets

```{r}
fc1 <-  "~/scratch-data/neta-poc/MSigDB/c1.all.v6.0.entrez.gmt"
c1 <- msig_import(fc1)
#summary(c1)
```

### C2 Reactome


```{r}
fc2r <-  "~/scratch-data/neta-poc/MSigDB/c2.cp.reactome.v6.0.entrez.gmt"
c2r <- msig_import(fc2r)
#summary(c1)
```


### C4 Cancer Gene Neighborhood

```{r}
fc4_cgn <-  "~/scratch-data/neta-poc/MSigDB/c4.cgn.v6.0.entrez.gmt"
c4cgn <- msig_import(fc4_cgn)
#summary(c4cgn)
```

### C4 Cancer Modules

```{r}
fc4_cm <-  "~/scratch-data/neta-poc/MSigDB/c4.cm.v6.0.entrez.gmt"
c4cm <- msig_import(fc4_cm)
#summary(c4cgn)
```

### C6 Oncogenic signatures

```{r}
fc6 <-  "~/scratch-data/neta-poc/MSigDB/c6.all.v6.0.entrez.gmt"
c6 <- msig_import(fc6)
#summary(c6)
```

## Map Annotations

Convert the Boot.Vote network into an *igraph* object and map the annotations onto the network. 

```{r}
library(spacemap)
ig <- spacemap::adj2igraph(yy = net$yy, xy = net$xy, yinfo = yinfo, xinfo = xinfo)
```

If we query the attribute names of the nodes in the graph, we notice that the columns of `xinfo` and `yinfo` have been applied. 

```{r}
vertex_attr_names(graph = ig)
```

The igraph package has a number of ways to access the annotation information. For example, if we wish to confirm the chromosome location of GRB7, we can easily query: 

```{r}
vertex_attr(graph = ig, name = "chr", index = V(ig)[alias %in% "GRB7"])
```

## Hub Analysis

We first prioritize the CNA- and protein- hubs. If  the `bdeg` argument is specified, then we rank the hubs according to the average degree rank. Accordingly, the most highly ranked hubs will have the most consistently high degree across network ensemble. 

To rank the protein nodes, use the `rankHub` command and specify the `level = "y"` argument. 

```{r}
ig <- rankHub(ig = ig, bdeg = bdeg$yy, level = "y")
```

To rank the CNA nodes, specify the `level = "x"` argument. 

```{r}
ig <- rankHub(ig = ig, bdeg = bdeg$xy, level = "x")
```

### Identify cis and trans

Next label $x-y$ edges as being regulated in cis or in trans. The [GenomicRanges](http://bioconductor.org/packages/release/bioc/html/GenomicRanges.html) R package and the genomic coordinates `chr,start,end` columns of `xinfo` and `yinfo` are required for this step.  Now we can label the $x-y$ edges with either "cis" or "trans" in the `cis_trans` edge attribute of `ig`.

```{r}
ig <- cisTrans(ig = ig, level = "x-y", cw = 2e6)
```

### Report 

```{r}
xhubs <- reportHubs(ig, top = 600, level = "x")
saveRDS(xhubs, "~/scratch-data/neta-poc/neta/protein-boot-vote-xhubs.rds")
saveRDS(xhubs[order(rownames(xhubs)),], "~/scratch-data/neta-poc/neta/protein-boot-vote-xhubs-sorted-by-id.rds")
chr_start <- sapply(strsplit(x = rownames(xhubs), split = "_"), function(x) x[2])
chr_end <- sapply(strsplit(x = rownames(xhubs), split = "_"), function(x) x[2])
#xhubs$`Potential # cis` <- as.character(xhubs$`Potential # cis`)
#cis_len <- sapply(strsplit(xhubs$`cis genes`,","), length)
xhubdir <- "~/scratch-data/neta-poc/neta/tables/xhubs"
library(xtable)
xhubstab <- cbind(Cytoband = xhubs$hub, Start = chr_start, End = chr_end, xhubs[,c(2,3,4)])
head(xhubstab)
#Part of table  in section ????
cap <- "Prioritized CNA hubs of \texttt{spaceMap poc-net} in each row with the number of -cis and -trans edges emanating from CNA hub, potential number of -cis relationships for the Hub, and which genes are in -cis with the CNA hub."
print(xtable(xhubstab, display  = c("d","s", "d", "d", "s", "d", "s"), caption = cap),
      include.rownames=FALSE,
      type = "latex", 
      file = file.path(xhubdir, "poc-spacemap-cna-hub.tex"))
#excel file
# xlsx::write.xlsx(x = xhubtab, file = file.path(netadir, "spacemap-xhub-prot.xlsx"))
# #supplementary section 
# capsup <- "Prioritized CNA hubs of \texttt{spaceMap prot-net} in each row with gene coordinates, the number of -cis and -trans edges emanating from the CNA hub, potential number of -cis relationships for the Hub, which genes are in -cis with the CNA hub, and the mean (std. deviation) rank of CNA degree across 1000 bootstrap replications."
# print(xtable(xhubtab[,c(1,3:8)], display  = c("d","s", "d", "d","s", "d", "s", "s"), caption = capsup),
#       include.rownames=FALSE,
#       type = "latex", 
#       file = file.path(netadir, "spacemap-cna-hub-with-prot-supp.tex"))
```

```{r, echo=FALSE}
kable(xhubs, row.names = FALSE)
```

Similarly, we can report the top 10 protein hubs,  their degrees in the  final network, and a description of each hub, if the `description` column was specified in `yinfo`. 

```{r}
yhubs <- reportHubs(ig, top = 10, level = "y")
```

```{r, echo=FALSE}
kable(yhubs, row.names = FALSE)
```

### GO-neighbor percentage

```{r}
hgp <- xHubEnrich(ig = ig, go2eg = go2eg)
#order by degree and GO-neighbor percentage
hgp2 <- hgp[order(hgp$degree, hgp$neighbor_percentage, decreasing = T)[1:30],]
hgp3 <- hgp2[hgp2$neighbor_percentage > 40,]
```

```{r, echo=FALSE}
kable(hgp3, row.names = FALSE)
```

## Module Analysis

There are many criteria to define modules of a network. This toolkit  allows users to import the module membership information by themselves (see `mods` argument in [modEnrich](https://topherconley.github.io/spacemap/reference/modEnrich.html)).  

In the spaceMap publication, we use the edge-betweenness algorithm in *igraph*.

```{r, eval = F}
library(igraph)
mods <- cluster_edge_betweenness(ig)
saveRDS(mods, file = "~/scratch-data/neta-poc/neta/cluster_edge_betweenness.rds")
```

```{r, echo = F}
mods <- readRDS(file = "~/scratch-data/neta-poc/neta/cluster_edge_betweenness.rds")
```



```{r}
write_tex <- function(etab, path, cap = "") { 
  library(xtable)
  #Part of table  in section ????
  print(xtable(xhubstab, display  = c("d","s", "s", "s", "s"), caption = cap),
        include.rownames=FALSE,
        type = "latex", 
        file = file.path(xhubdir, path))
}
```


```{r}
c1_mods <- modEnrich(ig = ig, mods = mods, levels = "y", go2eg = c1,  prefix = "C1:")
enrichdir <- "~/scratch-data/neta-poc/neta/tables/enrichment/"

write_tex(etab = c1_mods$etab, path = file.path(enrichdir, "prot-c1.rds"))

str(c1_mods$etab)
c4cgn_mods <- modEnrich(ig = c1_mods$ig, mods = mods, levels = "y", go2eg = c4cgn, prefix = "C4,cgn:")
#c4cgn_mods$etab
c4cm_mods <- modEnrich(ig = c4cgn_mods$ig, mods = mods, levels = "y", go2eg = c4cm, prefix = "C4,cm:")
#c4cm_mods$etab
c6_mods <- modEnrich(ig = c4cm_mods$ig, mods = mods, levels = "y", go2eg = c6, prefix = "C6:")
#c6_mods$etab
bioc_mods <- modEnrich(ig = c6_mods$ig, mods = mods, levels = "y", go2eg = go2eg, prefix = "GO:", 
                       process_alias = process_alias)
c2r_mods <- modEnrich(ig = bioc_mods$ig, mods = mods, levels = "y", go2eg = c2r,  prefix = "C2,r:")
#c2r_mods$etab
bioc_mods$etab
#knitr::kable(bioc_mods$etab, row.names = F)
all_mods <- c2r_mods
saveRDS(all_mods, file = "~/scratch-data/neta-poc/neta/all_170914_mods.rds")
all_mods <- readRDS(file = "~/scratch-data/neta-poc/neta/all_170914_mods.rds")
#c5mods <- modEnrich(ig = ig, mods = mods, levels = "y", go2eg = c5, prefix = "C5:")
#saveRDS(c5mods, file = "~/scratch-data/neta-poc/neta/c5mods.rds")
```

```{r}
msigmods <- do.call(rbind, list(c4cm_mods$etab, c4cgn_mods$etab,
                                c6_mods$etab, bioc_mods$etab))
back_fill <- function(x) {
  fill <- x[1]
  for (i in seq_along(x)) {
    if (nchar(x[i]) == 0) {
      x[i] <- fill
    } else {
      fill <- x[i]
    }
  }
  x
}
msigmods$`Module (size)` <- back_fill(msigmods$`Module (size)`)
idmod <- sapply(strsplit(x = msigmods$`Module (size)`, split = ":"), function(x) x[2])
cidmod <- sapply(strsplit(x = msigmods$`Module (size)`, split = ":"), function(x) x[1])
msigmods <- msigmods[order(idmod, cidmod),]
kable(msigmods)
```

## Export for Visualization

```{r}
filename <- "~/scratch-data/neta-poc/neta/all-poc-prot-boot-vote.graphml"
#delete nodes without edges from the graph object
vis <- delete_vertices(graph = all_mods$ig, v = V(all_mods$ig)[igraph::degree(all_mods$ig) == 0])
igraph::write_graph(graph = vis, file = filename, format = "graphml")
```



```{r}
write.table(x = cbind(alias = V(all_mods$ig)$alias, regualtes_in_cis = V(all_mods$ig)$regulates_in_cis), file = "~/scratch-data/neta-poc/neta/cis.txt")
```

```{r}
# filename <- "~/scratch-data/neta-poc/neta/c1-poc-prot-boot-vote.graphml"
# #delete nodes without edges from the graph object
# vis <- delete_vertices(graph = c1mods$ig, v = V(c1mods$ig)[igraph::degree(c1mods$ig) == 0])
# igraph::write_graph(graph = vis, file = filename, format = "graphml")
```


```{r}
# filename <- "~/scratch-data/neta-poc/neta/c4cgn-poc-prot-boot-vote.graphml"
# #delete nodes without edges from the graph object
# vis <- delete_vertices(graph = c4cgnmods$ig, v = V(c4cgnmods$ig)[igraph::degree(c4cgnmods$ig) == 0])
# igraph::write_graph(graph = vis, file = filename, format = "graphml")
```


```{r}
# filename <- "~/scratch-data/neta-poc/neta/c6-poc-prot-boot-vote.graphml"
# #delete nodes without edges from the graph object
# vis <- delete_vertices(graph = c6mods$ig, v = V(c6mods$ig)[igraph::degree(c6mods$ig) == 0])
# igraph::write_graph(graph = vis, file = filename, format = "graphml")
```


